File: python_hyper (2)/python_hyper/python_hyper/simplex_hull.py
import os
import sys
import argparse
import time
import numpy
import numpy as np
import scipy

import format_loader
from trimesh import TriMesh
from scipy.spatial import ConvexHull
import glob
from space_mapper import SpaceMapper

########################################
# CMD-line tool for getting filenames. #
########################################
if __name__ == '__main__':
	'''
	Uses ArgumentParser to parse the command line arguments.
	Input:
		parser - a precreated parser (If parser is None, creates a new parser)
	Outputs:
		Returns the arguments as a tuple in the following order:
			(in_mesh, Ts, Tmat)
	'''


	per_vertex_folder = "./cubeOut"
	in_transformations = glob.glob(per_vertex_folder + "/*.DMAT")
	print(len(in_transformations))
	in_transformations.sort()
	Ts = numpy.array([ format_loader.load_DMAT(transform_path).T for transform_path in in_transformations ])	
	num_poses = Ts.shape[0]
	num_verts = Ts.shape[1]
	Ts = numpy.swapaxes(Ts,0,1)
	Ts = Ts.reshape(num_verts, -1)
	
	
	
	root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
	numpy.set_printoptions(precision=4, suppress=True)
	import os,sys
	import mves2
	
	startTime = time.time()
	
	all_Ts = Ts.copy()
	## results stores: volume, solution, iter_num
	
	

	
	Ts_mapper = SpaceMapper.Uncorrellated_Space( Ts, dimension = 3 )
	uncorrelated = Ts_mapper.project( Ts )
	print( "uncorrelated data shape" )
	print( uncorrelated.shape )
	solution, weights, iter_num = mves2.MVES( uncorrelated, method = "qp")
	volume = abs( np.linalg.det( solution ) )
	print(solution.shape)

	print( "=> Best simplex found with volume:", volume )

	
	running_time = (time.time() - startTime)/60
	
	## If we have 4 bones, we can visualize the results in 3D!
	save_json = solution.shape[1] == 4
	if save_json:
		import json
		assert solution.shape[1] == 4
		simplex_vs = solution.T[:,:-1]
		norm_min = simplex_vs.min(axis=0)
		norm_scale = 1./( simplex_vs.max(axis=0) - simplex_vs.min(axis=0) ).max()
		def normalizer( pts ):
			return ( pts - norm_min ) * norm_scale
		
		out_vs = normalizer( uncorrelated )
		simplex_vs = normalizer( simplex_vs )*255.0
		json.dump( { "float_colors": out_vs.tolist() }, open("data.json",'w') )
		json.dump( { "vs": simplex_vs.tolist(), "faces": simplex_vs[ numpy.array([[0,3,2],[3,1,2],[3,0,1],[1,0,2]]) ].tolist() }, open("data-overlay.json",'w') )
	
	recovered = Ts_mapper.unproject( solution[:-1].T )

	
	output_path = os.path.join(per_vertex_folder, "result.txt")
	# if args.output is not None:
	output_path ='./cubeOut/result.txt'
	print( "Saving recovered results to:", output_path )
	format_loader.write_result(output_path, recovered, weights, iter_num, running_time, col_major=True)

File: cpp_hyper/cpp_hyper/simplexhull.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <stdexcept>
#include <cmath>
#include <algorithm>
#include <chrono>
#include <Eigen/Dense>

// PLACEHOLDERS for modules you must implement:
namespace format_loader {
    Eigen::MatrixXd load_DMAT(const std::string &path) {
        // TODO: Implement DMAT loading logic. Return Eigen::MatrixXd
        // This must mimic Python's format_loader.load_DMAT
        throw std::runtime_error("format_loader::load_DMAT not implemented.");
    }

    void write_result(const std::string &path, const Eigen::MatrixXd &res, const Eigen::MatrixXd &weights,
                      int iter_num, double running_time, bool col_major=true) {
        // TODO: Implement writing result logic as per Python code
        // Possibly write transforms, weights, iteration count, etc.
        std::cout << "write_result called with path:" << path << "\n";
    }
}

struct TriMesh {
    std::vector<Eigen::Vector3d> vs;
    static TriMesh FromOBJ_FileName(const std::string &fname) {
        // TODO: Implement or skip if not needed
        TriMesh mesh;
        return mesh;
    }
};

// Placeholder for SpaceMapper (PCA)
struct SpaceMapper {
    Eigen::RowVectorXd Xavg_;
    Eigen::MatrixXd V_;
    // Uncorrellated_Space returns a mapper with dimension=H
    static SpaceMapper Uncorrellated_Space(const Eigen::MatrixXd &X, bool enable_scale=true, double threshold=1e-6, int dimension=-1) {
        // TODO: Implement PCA-like logic
        SpaceMapper sm;
        // Compute mean, do SVD for PCA:
        Eigen::RowVectorXd mean = X.colwise().mean();
        sm.Xavg_ = mean;
        // center X:
        Eigen::MatrixXd centered = X.rowwise() - mean;
        Eigen::BDCSVD<Eigen::MatrixXd> svd(centered, Eigen::ComputeThinU|Eigen::ComputeThinV);
        if(dimension>0 && dimension<=X.cols()) {
            sm.V_ = svd.matrixV().leftCols(dimension-1); // dimension-1 as in python code
        } else {
            sm.V_ = svd.matrixV();
        }
        return sm;
    }

    Eigen::MatrixXd project(const Eigen::MatrixXd &X) const {
        // project: (X - Xavg)*V
        Eigen::MatrixXd centered = X.rowwise()-Xavg_;
        return centered*V_;
    }

    Eigen::MatrixXd unproject(const Eigen::MatrixXd &Y) const {
        // unproject: Y * V'. plus Xavg
        return Y*V_.transpose()+Xavg_;
    }
};

// Placeholder for MVES function in mves2
namespace mves2 {
    std::tuple<Eigen::MatrixXd,Eigen::MatrixXd,int> MVES(const Eigen::MatrixXd &data, const std::string &method="qp") {
        // TODO: Implement MVES or stable simplex finding logic
        // Return (solution, weights, iter_num)
        // solution: (dim x (dim+1)) matrix defining simplex
        // weights: 
        // iter_num: number of iterations
        // For now, return a dummy simplex and weights
        int dim = data.cols();
        Eigen::MatrixXd solution(dim, dim+1);
        // create a simplex by using standard unit vectors and zero
        // e.g. solution = [0,...; e1; e2; ...]
        // Just fill with identity and zero:
        solution.setZero();
        for (int i=0;i<dim;i++){
            solution(i,i)=1.0;
        }
        Eigen::MatrixXd weights = Eigen::MatrixXd::Ones(data.rows(),1);
        int iter_num=1;
        return std::make_tuple(solution, weights, iter_num);
    }
}

int main(){
    std::string per_vertex_folder = "./cubeOut";
    // glob DMAT files:
    // C++17 <filesystem> can be used:
    std::vector<std::string> in_transformations;
    // TODO: fill in_transformations with all *.DMAT in per_vertex_folder
    // For simplicity, assume we have them:
    // in_transformations = ...

    std::cout<<"Found "<<in_transformations.size()<<" DMAT files.\n";
    std::sort(in_transformations.begin(), in_transformations.end());

    // Ts: load all DMAT, shape: (#files, ???)
    // Python: Ts = np.array([ format_loader.load_DMAT(p).T for p in in_transformations ])
    // Suppose each DMAT is (something?), we must guess shapes.
    // In python code: Ts is #poses x #verts x ?
    // Then Ts swapped axes and reshaped. Without actual data we can guess:
    int num_poses = (int)in_transformations.size();
    Eigen::MatrixXd first = format_loader::load_DMAT(in_transformations[0]);
    // shape of first?
    // In python: first.T means we must also transpose after load:
    Eigen::MatrixXd firstT = first.transpose();
    int verts = firstT.rows();
    int cols = firstT.cols();
    // Ts shape after loading all:
    // Ts: (num_poses, verts, cols)
    // We'll store them in a big 3D structure or just vector:
    std::vector<Eigen::MatrixXd> Ts_list(num_poses);
    Ts_list[0]=firstT;
    for (int i=1;i<num_poses;i++){
        Eigen::MatrixXd M = format_loader::load_DMAT(in_transformations[i]);
        M=M.transpose();
        if(M.rows()!=verts || M.cols()!=cols) {
            throw std::runtime_error("All DMAT must have same shape");
        }
        Ts_list[i]=M;
    }

    // Ts shape: (num_poses, verts, cols)
    // python: Ts = np.swapaxes(Ts,0,1)
    // now Ts: (verts, num_poses, cols)
    // then reshape to (verts, -1)
    // The final Ts is (verts, num_poses*cols)
    // We'll create a big matrix Ts (verts, num_poses*cols)
    Eigen::MatrixXd Ts(verts, num_poses*cols);
    for(int v=0; v<verts; v++){
        for(int p=0; p<num_poses; p++){
            for(int c=0;c<cols;c++){
                Ts(v, p*cols+c) = Ts_list[p](v,c);
            }
        }
    }

    // no upper code with triMesh and rest_??? It's incomplete:
    // The python code references rest_mesh and per_vertex transformations,
    // Here we just replicate logic:

    // Suppose we do something similar to python code:
    // Create a SpaceMapper and do PCA dimension=3
    SpaceMapper Ts_mapper = SpaceMapper::Uncorrellated_Space(Ts, true,1e-6,3);
    Eigen::MatrixXd uncorrelated = Ts_mapper.project(Ts);
    std::cout<<"uncorrelated data shape: "<<uncorrelated.rows()<<"x"<<uncorrelated.cols()<<"\n";

    // mves2.MVES
    Eigen::MatrixXd solution, weights;
    int iter_num;
    std::tie(solution, weights, iter_num) = mves2::MVES(uncorrelated,"qp");

    double volume = std::abs(solution.determinant());
    std::cout<<"=> Best simplex found with volume: "<<volume<<"\n";

    double running_time=0.0; // measure time if needed

    // if solution has dimension h=3, solution shape (2 x ???)
    // If solution.shape[1]==4 in python means dimension +1=4 => dimension=3
    // if (solution.cols()==4) do save json:
    if(solution.cols()==4) {
        // normalizer etc:
        // We skip json output in C++
        // just print a message
        std::cout<<"Would save JSON here.\n";
    }

    Eigen::MatrixXd recovered = Ts_mapper.unproject(solution.topRows(solution.rows()-1).transpose());

    std::string output_path = per_vertex_folder + "/result.txt";
    std::cout<<"Saving recovered results to:"<<output_path<<"\n";
    format_loader::write_result(output_path, recovered, weights, iter_num, running_time, true);

    return 0;
}